lsp/gitlab_duo.lua

```diff
@@ -0,0 +1,473 @@
+    status = tonumber(output:match('\n(%d+)$')) or 0
+  else
+    body = output
+  end
+
+  return {
+    status = status,
+    body = body,
+  }
+end
+
+-- Token management
+local function save_token(token_data)
+  token_data.saved_at = os.time()
+  local file = io.open(config.token_file, 'w')
+  if file then
+    file:write(vim.json.encode(token_data))
+    file:close()
+    return true
+  end
+  return false
+end
+
+local function load_token()
+  local file = io.open(config.token_file, 'r')
+  if file then
+    local content = file:read('*all')
+    file:close()
+    local ok, token_data = pcall(vim.json.decode, content)
+    if ok then
+      return token_data
+    end
+  end
+  return nil
+end
+
+local function is_token_expired(token_data)
+  if not token_data or not token_data.saved_at or not token_data.expires_in then
+    return true
+  end
+  local token_age = os.time() - token_data.saved_at
+  return token_age >= (token_data.expires_in - 60) -- 60 second buffer
+end
+
+local function refresh_access_token(refresh_token)
+  vim.notify('Refreshing GitLab OAuth token...', vim.log.levels.INFO)
+
+  local response = curl_post(
+    config.gitlab_url .. '/oauth/token',
+    string.format('client_id=%s&refresh_token=%s&grant_type=refresh_token', config.client_id, refresh_token),
+    { ['Content-Type'] = 'application/x-www-form-urlencoded' }
+  )
+
+  if response.status ~= 200 then
+    vim.notify('Failed to refresh token: ' .. (response.body or 'Unknown error'), vim.log.levels.ERROR)
+    return nil
+  end
+
+  local ok, body = pcall(vim.json.decode, response.body)
+  if not ok or not body.access_token then
+    vim.notify('Invalid refresh response', vim.log.levels.ERROR)
+    return nil
+  end
+
+  save_token(body)
+  vim.notify('Token refreshed successfully', vim.log.levels.INFO)
+  return body
+end
+
+local function get_valid_token()
+  local token_data = load_token()
+
+  if not token_data then
+    return nil, 'no_token'
+  end
+
+  if is_token_expired(token_data) then
+    if token_data.refresh_token then
+      local new_token_data = refresh_access_token(token_data.refresh_token)
+      if new_token_data then
+        return new_token_data.access_token, 'refreshed'
+      end
+      return nil, 'refresh_failed'
+    end
+    return nil, 'expired'
+  end
+
+  return token_data.access_token, 'valid'
+end
+
+-- OAuth Device Flow
+local function device_authorization()
+  local response = curl_post(
+    config.gitlab_url .. '/oauth/authorize_device',
+    string.format('client_id=%s&scope=%s', config.client_id, config.scopes),
+    { ['Content-Type'] = 'application/x-www-form-urlencoded' }
+  )
+
+  if response.status ~= 200 then
+    vim.notify('Device authorization failed: ' .. response.status, vim.log.levels.ERROR)
+    return nil
+  end
+
+  local ok, data = pcall(vim.json.decode, response.body)
+  if not ok then
+    vim.notify('Failed to parse authorization response', vim.log.levels.ERROR)
+    return nil
+  end
```

Comment:
might be better just to drop the `pcall` so that an actual error is thrown with a stacktrace.